ormolu-live/cbits/init.c

```diff
@@ -1,14 +1,28 @@
+#include "Rts.h"
+
 #include "Main_stub.h"
-#include <Rts.h>
+
+void malloc_inspect_all(void (*handler)(void *start, void *end,
+                                        size_t used_bytes, void *callback_arg),
+                        void *arg);
+
+static void malloc_inspect_all_handler(void *start, void *end,
+                                       size_t used_bytes, void *callback_arg) {
+  if (used_bytes == 0) {
+    memset(start, 0, (size_t)end - (size_t)start);
+  }
+}
 
 __attribute__((export_name("wizer.initialize"))) void __wizer_initialize(void) {
   char *args[] = {
-      "ormolu-live.wasm", "+RTS", "--nonmoving-gc", "-H64m", "-RTS", NULL};
+      "ormolu-live.wasm", "+RTS", "-H64m", "-RTS", NULL};
   int argc = sizeof(args) / sizeof(args[0]) - 1;
   char **argv = args;
   hs_init_with_rtsopts(&argc, &argv);
   evaluateFixityInfo();
   hs_perform_gc();
   hs_perform_gc();
   rts_clearMemory();
+  malloc_inspect_all(malloc_inspect_all_handler, NULL);
+  memset(0, 0, (size_t)__builtin_frame_address(0));
```

Comment:
Is this line necessary for correctness, or is it another optimization? Size-wise, it seems to cause a reduction by ~4KB from 18.596MB to 18.592MB.

Reply:
You mean the `memset` line? It's meant to wipe the stack region which starts from `0x0` (because ghc is configured to link with `-Wl,--stack-first`) up to the current stack pointer. It's an independent optimization, suitable to be done just before exiting the wizer initialize function.